KeychainMinderAgent/KeychainMinderAgent.m

```diff
@@ -0,0 +1,87 @@
+/// Copyright 2015 Google Inc. All rights reserved.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///    http://www.apache.org/licenses/LICENSE-2.0
+///
+///    Unless required by applicable law or agreed to in writing, software
+///    distributed under the License is distributed on an "AS IS" BASIS,
+///    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+///    See the License for the specific language governing permissions and
+///    limitations under the License.
+
+#import "KeychainMinderAgent.h"
+#import <MOLCertificate/MOLCertificate.h>
+#import <MOLCodesignChecker/MOLCodesignChecker.h>
+
+@interface KeychainMinderAgent()
+
+@property (nonatomic, strong) NSXPCListener *listener;
+@property (nonatomic, strong) NSMutableData *passwordData;
+
+@end
+
+@implementation KeychainMinderAgent
+
+- (id)init {
+  self = [super init];
+  if (self != nil) {
+    // Set up our XPC listener to handle requests on our Mach service.
+    _listener = [[NSXPCListener alloc] initWithMachServiceName:kKeychainMinderAgentMachServiceName];
+    _listener.delegate = self;
+  }
+  return self;
+}
+
+- (void)run
+{
+  // Tell the XPC listener to start processing requests.
+  [self.listener resume];
+  
+  // Run the run loop forever.
+  [[NSRunLoop currentRunLoop] run];
+}
+
+- (BOOL)listener:(NSXPCListener *)listener shouldAcceptNewConnection:(NSXPCConnection *)newConn {
+  pid_t pid = newConn.processIdentifier;
+  
+  MOLCodesignChecker *selfCS = [[MOLCodesignChecker alloc] initWithSelf];
+  MOLCodesignChecker *remoteCS = [[MOLCodesignChecker alloc] initWithPID:pid];
+  
+  // Add an exemption for authorizationhost.bundle connections. This is needed so the authorization
+  // plugin use this XPC service.
+  // For testing
+  // NSString *const ahReqString = @"identifier \"com.apple.xctest\" and anchor apple";
```

Comment:
You can use this to determine if currently being tested and add the xctest requirement as appropriate:

```objc
+ (BOOL)beingUnitTested {
  BOOL answer = NO;
  Class testProbeClass = NSClassFromString(@"XCTestProbe");
  if (testProbeClass != Nil) {
    SEL selector = NSSelectorFromString(@"isTesting");
    NSMethodSignature *sig = [testProbeClass methodSignatureForSelector:selector];
    NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:sig];
    [invocation setSelector:selector];
    [invocation invokeWithTarget:testProbeClass];
    [invocation getReturnValue:&answer];
  }
  return answer;
}
```